### PR TITLE
Using asio::io_context::strands instead of locks

### PR DESCRIPTION
**Pull request by graetzer**
- Upgrade boost::asio by using the latest standalone version
- Replace all socket locks with strands

http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/403/

### CODE REVIEW

**Comment by mpoeter** [comment]
This is a proper fix for the issue I described in #5139 / #5141 
VTune results look much better now (no significant contention on the strands' mutexes). In my simple test scenarios with arangobench the number of created threads is much lower and the throughput increases by ~20%.

File: lib/SimpleHttpClient/Communicator.cpp [hunk 1]
```diff
@@ -140,6 +140,8 @@ Communicator::Communicator() : _curl(nullptr), _mc(CURLM_OK), _enabled(true) {
   if (_curl == nullptr) {
     THROW_ARANGO_EXCEPTION_MESSAGE(TRI_ERROR_OUT_OF_MEMORY, "unable to initialize curl");
   }
+  
+  curl_multi_setopt(_curl, CURLMOPT_PIPELINING, 0);
```
Comment (graetzer) [inline lib/SimpleHttpClient/Communicator.cpp]: test

File: arangod/GeneralServer/GeneralServer.cpp [hunk 2]
```diff
@@ -93,8 +93,9 @@ bool GeneralServer::openEndpoint(Endpoint* endpoint) {
     protocolType = ProtocolType::HTTP;
   }
 
-  std::unique_ptr<ListenTask> task(new GeneralListenTask(
-      SchedulerFeature::SCHEDULER->eventLoop(), this, endpoint, protocolType));
+  std::unique_ptr<ListenTask> task;
+  task.reset(new GeneralListenTask(SchedulerFeature::SCHEDULER->eventLoop(),
+                                   this, endpoint, protocolType));
```
[the remark this answers is not in the document]
Comment (graetzer) [reply arangod/GeneralServer/GeneralServer.cpp]: that did not compile on clang

[the remark this answers is not in the document]
Comment (mchacki) [reply arangod/GeneralServer/GeneralServer.cpp]: ok than :+1:

**Pull request merged**